// next-sitemap/dist/cjs/url/util/index.js
"use strict";
/* eslint-disable no-useless-escape */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDefaultLocaleReplace = exports.isNextInternalUrl = exports.generateUrl = exports.isURL = exports.cleanPath = void 0;
exports.cleanPath = (text) => {
    return text.replace(/([^:])(\/\/+)/g, '$1/');
};
exports.isURL = (text) => {
    // old: /^(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&'\(\)\*\+,;=.]+$/
    return /^https?:\/\//i.test(text);
};
exports.generateUrl = (baseUrl, slug) => {
    return exports.isURL(slug) ? exports.cleanPath(slug) : exports.cleanPath(`${baseUrl}/${slug}`);
};
/**
 * Checks whether a url is next.js specific or not
 * @param path path check
 */
exports.isNextInternalUrl = (path) => {
    return new RegExp(/[^\/]*^.[_]|^\/404$|(?:\[)/g).test(path);
};
/**
 * Creates a replace function to replace the default locale
 * Avoids creating the same RegExp within each replace
 *
 * Replaces only if the path does not contain the locale as an actual valid path
 *
 * Given a default locale of en-US it replaces:
 * /en-US -> /
 * /en-US/home -> /home
 * /en-US/home/ -> /home/
 *
 * Does not replace if its actual page
 * /en-USA -> /en-USA
 * /en-USA/home -> /en-USA/home
 * /en-US-home -> /en-US-home
 *
 * @param defaultLocale defaultLocale as provided by i18n within next config
 */
exports.createDefaultLocaleReplace = (defaultLocale) => {
    const defaultLocaleRegExp = new RegExp(`^/${defaultLocale}($|/)`);
    return (path) => path.replace(defaultLocaleRegExp, '/');
};
